farmerbot/parser/validator.go

```diff
@@ -0,0 +1,86 @@
+package parser
+
+import (
+	"fmt"
+	"slices"
+
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/internal"
+)
+
+// wrapper for validateInput
+func ValidateConfig(input internal.Config, network string) error {
+	manager := substrate.NewManager(internal.SubstrateURLs[network]...)
+	subConn, err := manager.Substrate()
+	if err != nil {
+		return err
+	}
+	defer subConn.Close()
+	return validateInput(input, subConn)
+}
+
+// validateInput validates that included, excluded and priority nodes are in the farm
+func validateInput(input internal.Config, sub internal.Substrate) error {
+	nodes, err := sub.GetNodes(input.FarmID)
+	if err != nil {
+		return fmt.Errorf("couldn't retrieve node for %d : %v", input.FarmID, err)
+	}
+	nodesMap := make(map[uint32]bool)
+	for _, node := range nodes {
+		nodesMap[node] = true
+	}
+	includedNodes := make(map[uint32]bool)
+	if len(input.IncludedNodes) != 0 {
+		if err := validateIncludedNodes(input.IncludedNodes, input.ExcludedNodes, nodesMap); err != nil {
+			return err
+		}
+		for _, includedNode := range input.IncludedNodes {
+			includedNodes[includedNode] = true
+		}
+	} else {
+		includedNodes = nodesMap
+	}
+	if err := validateExcludedNodes(input.ExcludedNodes, nodesMap); err != nil {
+		return err
+	}
+	if err := validatePriorityOrNeverShutdown("priority", input.PriorityNodes, input.ExcludedNodes, includedNodes); err != nil {
+		return err
+	}
+	if err := validatePriorityOrNeverShutdown("never shutdown", input.NeverShutDownNodes, input.ExcludedNodes, includedNodes); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateIncludedNodes(included, excluded []uint32, farmNodes map[uint32]bool) error {
+	for _, node := range included {
+		if _, ok := farmNodes[node]; !ok {
+			return fmt.Errorf("included node with id %d doesn't exist in the farm", node)
+		}
+		if slices.Contains(excluded, node) {
+			return fmt.Errorf("cannot include and exclude the same node %d", node)
+		}
+	}
+	return nil
+}
+
+func validatePriorityOrNeverShutdown(typeOfValidation string, toBeValidated, excluded []uint32, nodes map[uint32]bool) error {
+	for _, node := range toBeValidated {
+		if _, ok := nodes[node]; !ok {
+			return fmt.Errorf("%s node with id %d doesn't exist in the included nodes ", typeOfValidation, node)
+		}
+		if slices.Contains(excluded, node) {
+			return fmt.Errorf("node %d can not be both %s and excluded", node, typeOfValidation)
+		}
```

Comment:
at this point if the node is in the included nodes, it won't be in the excluded nodes, no?

Reply:
yes it will not be in the excluded, this check is added for the case of all nodes included because in that case there is no validation for included nodes so priority and nevershutdown nodes need to be validated against excluded nodes

Reply:
then you'd need to remove the excluded nodes from the list of included nodes, check line 41, included nodes should NOT be a map that include excluded nodes

Reply:
after that there is no need for this check right?

Reply:
yes